src/realm/mixed.cpp

```diff
@@ -335,8 +335,8 @@ int Mixed::compare(const Mixed& b) const noexcept
             if (type == type_TypeOfValue && b.get_type() == type_TypeOfValue) {
                 return TypeOfValue(int_val).matches(TypeOfValue(b.int_val)) ? 0 : compare_generic(int_val, b.int_val);
             }
-            if ((type == type_List || type == type_Dictionary) && type == b.get_type()) {
-                return compare_generic(int_val, b.int_val);
+            if ((type == type_List || type == type_Dictionary)) {
+                return m_type == b.m_type ? 0 : m_type < b.m_type ? -1 : 1;
```

Comment:
This is subtle, I did not catch during the previous code review, but it makes sense.